Backend refactor

eugene-eko2000 opened the pull request.

eldarm commented on api/endpoints.go:
```diff
@@ -0,0 +1,9 @@
+	GetMapEndpoint        = "/get_map"
+	ReadReferralEndpoint  = "/readreferral"
+	WriteReferralEndpoint = "/writereferral"
+)
```
If you indeed want a separate file and package for "api"s, then structures UserArgs, ReporArgs, etc. also belong here.

If you did so because of client.go / client/main.go than
1. Why do so if this was not necessary fro backend/main.go?
2. clint is esentially a test (just different level test), for tests such separation is normally not done in go.

eugene-eko2000 replied:
> If you indeed want a separate file and package for "api"s, then structures UserArgs, ReporArgs, etc. also belong here.

It's a good idea, I'm moving all API related structures into the API package. And yes, the client uses only API and never needs anything else from the backend.

eldarm commented on backend/impl/reeferral.go:
```diff
@@ -1,19 +1,12 @@
-package service
```
referral, not reEferral

eugene-eko2000 replied:
Sure, it's a typo. Done.

eldarm commented on backend/impl/db.go:
```diff
@@ -34,47 +40,36 @@ func validateResult(r sql.Result, e error) error {
+	db *sql.DB
+}
+
+func (s *sqlDB) updateUser(u UserArgs) error {
```
What is the value of
func (s *sqlDB) updateUser(u UserArgs) error
vs
func updateUser(s *sqlDB, u UserArgs) error
?

eugene-eko2000 replied:
I'd ask you about that, the function was introduced by you. I only added a receiver to it.

eugene-eko2000 replied:
TL;DR: Readability.

Compare:
db.updateUser(args)
vs
updateUser(db, args)
The former locates you to a db definition. The latter gives you no idea which updateUser you're calling and where to look it for, considering there is a function with the similar name.

eldarm commented on backend/impl/db.go:
```diff
@@ -34,47 +40,36 @@ func validateResult(r sql.Result, e error) error {
 }
 
-func updateUser(u UserArgs) error {
+type sqlDB struct {
```
If we don't make DB function methods, we won't need double wrapping of DB object. *sql.DB will be a member of handler type.

eugene-eko2000 replied:
I'd say it gives a good point about the function area. Gin handlers and db functions have very similar names. sqlDB receiver and handler receiver allow clearly separate them from each other. db.db isn't a big problem there.

eldarm reviewed (changes requested):
1. impl folder is clearly unnecessary. main.go sands out enough to make it unnecessary.
2. I use these scripts A LOT. So, if you like long script names like verylongScriptNameToBuildOurServer.sh, that's ok. Create script folder and put it there. But specially for me, please, keep b.sh, d.sh and c.sh, it's ok if they'll simply forward to yours.
3. I expected just merge of referrals and a single db connection. This change has much larger scale. It blocks me since many changes go itno renamed, moved or new files.

eugene-eko2000 commented on backend/impl/user.go:
```diff
@@ -13,7 +13,7 @@ type UserArgs struct {
 	Avatar  string `json:"avatar"`
 }
 
-func UpdateUser(c *gin.Context) {
```
That sounds reasonable, done.

eugene-eko2000 commented on backend/impl/server.go:
```diff
@@ -0,0 +1,50 @@
+	sDB *sqlDB
+}
+
+func newHandler(router *gin.Engine) (*handler, error) {
```
You're looking not at the latest commit in the PR. It's removed.

eugene-eko2000 commented on backend/impl/server.go:
```diff
@@ -0,0 +1,50 @@
+
+	handler, err := newHandler(router)
+	if err != nil {
+		log.Errorf("http handler creation: %w", err)
```
Renamed http handler -> gin handlers receiver.

eugene-eko2000 commented:
> 1. impl folder is clearly unnecessary. main.go sands out enough to make it unnecessary.

I swapped it, moved main into another dir instead.

> 2. I use these scripts A LOT. So, if you like long script names like verylongScriptNameToBuildOurServer.sh, that's ok. Create script folder and put it there. But specially for me, please, keep b.sh, d.sh and c.sh, it's ok if they'll simply forward to yours.

See my comment to the script. TL'DR: I prefer readable names. b.sh doesn't give me an idea of what it is.

> 3. I expected just merge of referrals and a single db connection. This change has much larger scale. It blocks me since many changes go itno renamed, moved or new files.

Sorry, I don't think I could avoid that. Most of changes are related to the db connection refactoring. Also I moved backend filed from impl dir back to backend dir. Hope that helps you.

eugene-eko2000 commented on backend/impl/db.go:
```diff
@@ -1,20 +1,26 @@
-package be
```
Not sure I'm following. If you're about Go naming that's a rule for package names: https://go.dev/doc/effective_go#package-names

... package name should be good: short, concise, evocative ... I believe "be" is neither concise nor evocative.

eugene-eko2000 commented on build_all.sh:
```diff
@@ -0,0 +1,9 @@
+go build -o bin/client client/main.go
+
+echo "Building MAC service..."
+GOARCH="arm64" GOOS="darwin" go build -o bin/serviceapp backend/main.go
```
I already mentioned that it turned out into a non-runnable binary. After removing dot the binary runs correctly. See the previous PR comment.

eugene-eko2000 commented on build_all.sh:
```diff
@@ -0,0 +1,9 @@
+echo "Building client..."
```
./bu<tab><enter> and you're all set. Same number of key presses, readability is much better. 
Also you can set your own aliases for these scripts. Let's keep the stuff readable in the repo and build custom ergonomics locally if necessary.
